src/backend/distributed/executor/partitioned_intermediate_results.c

```diff
@@ -0,0 +1,428 @@
+			[Anum_pg_dist_shard_shardmaxvalue - 1] = maxValues[partitionIndex]
+		};
+		bool nullsArray[Natts_pg_dist_shard] = {
+			[Anum_pg_dist_shard_shardminvalue - 1] = minValueNulls[partitionIndex],
+			[Anum_pg_dist_shard_shardmaxvalue - 1] = maxValueNulls[partitionIndex]
+		};
+
+		shardIntervalArray[partitionIndex] =
+			DeformedDistShardTupleToShardInterval(datumArray, nullsArray,
+												  partitionColumn->vartype,
+												  partitionColumn->vartypmod);
+		shardIntervalArray[partitionIndex]->shardIndex = partitionIndex;
+	}
+
+	DistTableCacheEntry *result = palloc0(sizeof(DistTableCacheEntry));
+	result->partitionMethod = partitionMethod;
+	result->partitionColumn = partitionColumn;
+	result->shardIntervalCompareFunction = shardIntervalCompare;
+	result->shardColumnCompareFunction = shardColumnCompare;
+	result->hashFunction = hashFunction;
+	result->sortedShardIntervalArray =
+		SortShardIntervalArray(shardIntervalArray, partitionCount,
+							   partitionColumn->varcollid, shardIntervalCompare);
+	result->hasUninitializedShardInterval =
+		HasUninitializedShardInterval(result->sortedShardIntervalArray, partitionCount);
+	result->hasOverlappingShardInterval =
+		result->hasUninitializedShardInterval ||
+		HasOverlappingShardInterval(result->sortedShardIntervalArray, partitionCount,
+									partitionColumn->varcollid, shardIntervalCompare);
+	ErrorIfInconsistentShardIntervals(result);
+
+	result->shardIntervalArrayLength = partitionCount;
+
+	return result;
+}
+
+
+static DestReceiver *
+CreatePartitionedResultDestReceiver(char *resultIdPrefix, int partitionColumnIndex,
+									int partitionCount, TupleDesc tupleDescriptor,
+									bool binaryCopy, DistTableCacheEntry *shardSearchInfo,
+									MemoryContext perTupleContext)
+{
+	PartitionedResultDestReceiver *resultDest =
+		palloc0(sizeof(PartitionedResultDestReceiver));
+
+	/* set up the DestReceiver function pointers */
+	resultDest->pub.receiveSlot = PartitionedResultDestReceiverReceive;
+	resultDest->pub.rStartup = PartitionedResultDestReceiverStartup;
+	resultDest->pub.rShutdown = PartitionedResultDestReceiverShutdown;
+	resultDest->pub.rDestroy = PartitionedResultDestReceiverDestroy;
+	resultDest->pub.mydest = DestCopyOut;
+
+	/* set up output parameters */
+	resultDest->resultIdPrefix = resultIdPrefix;
+	resultDest->perTupleContext = perTupleContext;
+	resultDest->partitionColumnIndex = partitionColumnIndex;
+	resultDest->partitionCount = partitionCount;
+	resultDest->shardSearchInfo = shardSearchInfo;
+	resultDest->tupleDescriptor = tupleDescriptor;
+	resultDest->binaryCopy = binaryCopy;
+	resultDest->partitionDestReceivers =
+		(DestReceiver **) palloc0(partitionCount * sizeof(DestReceiver *));
+
+	return (DestReceiver *) resultDest;
+}
+
+
+static void
+PartitionedResultDestReceiverStartup(DestReceiver *dest, int operation,
+									 TupleDesc inputTupleDescriptor)
+{
+	/* nothing to do here */
```

Comment:
Call `rStartup` on any non-null entries, since the interface is such that this could be coming from after a call to Shutdown

Reply:
> Is there a reason we don't want to create the files eagerly?

We can have a lot of partitions for range partitioned tables, and we don't want to lots of empty files if most of partitions are empty.